sei-protocol/sei-chain
# Revert "Revert "Custom mint module impl (#124)" (#202)"

This re-commits the mint module implementation. I think the previous issue @codchen saw was actually a race condition (https://discord.com/channels/@me/979184133676884039/1006798779774664785).
Here's what happened:
1. Tony runs loadtest ansible scripts which creates a genesis file without this change (i.e. using inflation). This happens sometime before 
```
commit 3ce1585500db6008025165d566f01eab18b4d3f1
Author: Philip Su <philip.su.522@gmail.com>
Date:   Tue Aug 9 21:15:44 2022 -0700

    Custom mint module impl (#124)
```
3. Philip merges the PR above at 21:15:44
4. Nodes with the new version begin coming up at 22:26:20 which fail:
```
TASK [start_sei : Wait for port to become open] ********************************
Tuesday 09 August 2022  22:26:20 -0700 (0:00:02.365)       0:10:16.533 ******** 
Tuesday 09 August 2022  22:26:20 -0700 (0:00:02.365)       0:10:16.533 ******** 
fatal: [sei-node-0 (testnet)]: FAILED! => changed=false 
  elapsed: 300
  msg: Timeout when waiting for 127.0.0.1:26657
``` 
The issue is basically that the genesis must be created with the correct version. I've verified via the following test:
1. create genesis without this PR, but don't start seid
2. install seid with this pr
3. start chain: 
```
panic: unknown field "inflation" in types.Minter

goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/codec.(*ProtoCodec).MustUnmarshalJSON(0x726f838?, {0xc0001b51e0?, 0x0?, 0x0?}, {0x63324b0?, 0xc0011880c0?})
```
I've also verified the following path works:
1. create genesis without this PR, start seid
2. install seid with this pr
3. resume chain

**Note** I added a new commit to migrate the params and tested by doing a chain upgrade. See [commit](https://github.com/sei-protocol/sei-chain/pull/217/files/5c5c35a8eeaf1a9e3906ee4da76c7bd062460487..6fdd8bfb284e48d52aa83cd51658de56ae0c9764) 
```
seid q mint params
genesis_epoch_provisions: "5000000.000000000000000000"
mint_denom: stake
reduction_factor: "0.500000000000000000"
reduction_period_in_epochs: "525600"
```
